File: frontend/src/app/core/services/collection.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { Collection, CollectionCreateDto, CollectionUpdateDto } from '../models/collection.model';
import { Memory } from '../models/memory.model';
import { PagedResponse } from '../models/api-response.model';

@Injectable({
  providedIn: 'root'
})
export class CollectionService {
  private readonly api = inject(ApiService);

  getCollections(): Observable<Collection[]> {
    return this.api.get<Collection[]>('/collections');
  }

  getCollectionById(id: string): Observable<Collection> {
    return this.api.get<Collection>(`/collections/${id}`);
  }

  createCollection(dto: CollectionCreateDto): Observable<Collection> {
    return this.api.post<Collection>('/collections', dto);
  }

  updateCollection(id: string, dto: CollectionUpdateDto): Observable<Collection> {
    return this.api.put<Collection>(`/collections/${id}`, dto);
  }

  deleteCollection(id: string): Observable<void> {
    return this.api.delete<void>(`/collections/${id}`);
  }

  getCollectionMemories(id: string, page = 0, size = 15): Observable<PagedResponse<Memory>> {
    return this.api.get<PagedResponse<Memory>>(`/collections/${id}/memories`, { page, size });
  }

  addMemoryToCollection(collectionId: string, memoryId: string): Observable<void> {
    return this.api.post<void>(`/collections/${collectionId}/memories/${memoryId}`, {});
  }

  removeMemoryFromCollection(collectionId: string, memoryId: string): Observable<void> {
    return this.api.delete<void>(`/collections/${collectionId}/memories/${memoryId}`);
  }
}

File: backend/src/main/java/com/memoryverse/service/CollectionService.java
package com.memoryverse.service;

import com.memoryverse.dto.request.CollectionCreateDto;
import com.memoryverse.dto.request.CollectionUpdateDto;
import com.memoryverse.dto.response.CollectionResponseDto;
import com.memoryverse.dto.response.MemoryResponseDto;
import com.memoryverse.dto.response.PagedResponse;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface CollectionService {

    List<CollectionResponseDto> getAllCollections();

    CollectionResponseDto getCollectionById(UUID id);

    CollectionResponseDto createCollection(CollectionCreateDto dto, UUID creatorId);

    CollectionResponseDto updateCollection(UUID id, CollectionUpdateDto dto, UUID currentUserId);

    void deleteCollection(UUID id, UUID currentUserId);

    void addMemoryToCollection(UUID collectionId, UUID memoryId, UUID currentUserId);

    void removeMemoryFromCollection(UUID collectionId, UUID memoryId, UUID currentUserId);

    PagedResponse<MemoryResponseDto> getCollectionMemories(UUID collectionId, Pageable pageable);
}
